File: python/src/utils/file_operations.py
import os
import json
import shutil
import re
from pathlib import Path
from typing import Any, Dict, List, Optional, Union, Tuple
from loguru import logger


class FileOperations:
    """File operations utilities for reading and writing files.
    
    This class provides methods for working with files and directories,
    including reading, writing, copying, and deleting files.
    """
    
    @staticmethod
    def read_file(file_path: Union[str, Path], encoding: str = "utf-8", diff_pattern: Optional[str] = None) -> Union[str, Tuple[str, bool]]:
        """Read a file and return its contents.
        
        Args:
            file_path: Path to the file to read
            encoding: File encoding
            diff_pattern: Optional git diff pattern to extract a specific section
                          Format: "<<<<<<< SEARCH\\n[content]\\n=======\\n[content]\\n>>>>>>> REPLACE"
            
        Returns:
            If diff_pattern is None: The file contents as a string
            If diff_pattern is provided: A tuple of (content, pattern_found)
            
        Raises:
            FileNotFoundError: If the file doesn't exist
            IOError: If there's an error reading the file
        """
        try:
            path = Path(file_path)
            logger.debug(f"Reading file: {path}")
            with open(path, "r", encoding=encoding) as f:
                content = f.read()
            
            if diff_pattern:
                # Extract content between SEARCH and ======= in the diff pattern
                search_pattern = r"<<<<<<< SEARCH\n(.*?)\n======="
                match = re.search(search_pattern, diff_pattern, re.DOTALL)
                
                if not match:
                    logger.warning(f"Invalid diff pattern format: {diff_pattern}")
                    return content, False
                
                search_content = match.group(1)
                pattern_found = search_content in content
                
                if pattern_found:
                    logger.debug(f"Found diff pattern in file: {path}")
                else:
                    logger.warning(f"Diff pattern not found in file: {path}")
                
                return content, pattern_found
            
            return content
        except FileNotFoundError:
            logger.error(f"File not found: {path}")
            raise
        except IOError as e:
            logger.error(f"Error reading file {path}: {str(e)}")
            raise
    
    @staticmethod
    def write_file(file_path: Union[str, Path], content: str, encoding: str = "utf-8", 
                  diff_pattern: Optional[str] = None) -> Union[None, bool]:
        """Write content to a file.
        
        Args:
            file_path: Path to the file to write
            content: Content to write to the file
            encoding: File encoding
            diff_pattern: Optional git diff pattern to replace a specific section
                          Format: "<<<<<<< SEARCH\\n[content to find]\\n=======\\n[content to replace with]\\n>>>>>>> REPLACE"
            
        Returns:
            If diff_pattern is provided: True if pattern was found and replaced, False otherwise
            If diff_pattern is None: None
            
        Raises:
            IOError: If there's an error writing the file
        """
        try:
            path = Path(file_path)
            # Create directory if it doesn't exist
            path.parent.mkdir(parents=True, exist_ok=True)
            
            if diff_pattern:
                # Parse the diff pattern
                search_pattern = r"<<<<<<< SEARCH\n(.*?)\n=======\n(.*?)\n>>>>>>> REPLACE"
                match = re.search(search_pattern, diff_pattern, re.DOTALL)
                
                if not match:
                    logger.warning(f"Invalid diff pattern format: {diff_pattern}")
                    return False
                
                search_content = match.group(1)
                replace_content = match.group(2)
                
                # Read existing content
                if path.exists():
                    with open(path, "r", encoding=encoding) as f:
                        file_content = f.read()
                else:
                    file_content = ""
                
                # Check if search content exists in the file
                if search_content not in file_content:
                    logger.warning(f"Search content not found in file: {path}")
                    return False
                
                # Replace the content
                new_content = file_content.replace(search_content, replace_content)
                
                logger.debug(f"Writing to file with diff pattern: {path}")
                with open(path, "w", encoding=encoding) as f:
                    f.write(new_content)
                
                return True
            else:
                logger.debug(f"Writing to file: {path}")
                with open(path, "w", encoding=encoding) as f:
                    f.write(content)
                
                return None
        except IOError as e:
            logger.error(f"Error writing to file {path}: {str(e)}")
            raise
    
    @staticmethod
    def append_to_file(file_path: Union[str, Path], content: str, encoding: str = "utf-8") -> None:
        """Append content to a file.
        
        Args:
            file_path: Path to the file to append to
            content: Content to append to the file
            encoding: File encoding
            
        Raises:
            IOError: If there's an error appending to the file
        """
        try:
            path = Path(file_path)
            # Create directory if it doesn't exist
            path.parent.mkdir(parents=True, exist_ok=True)
            
            logger.debug(f"Appending to file: {path}")
            with open(path, "a", encoding=encoding) as f:
                f.write(content)
        except IOError as e:
            logger.error(f"Error appending to file {path}: {str(e)}")
            raise
    
    @staticmethod
    def read_json(file_path: Union[str, Path]) -> Dict[str, Any]:
        """Read a JSON file and return its contents.
        
        Args:
            file_path: Path to the JSON file to read
            
        Returns:
            The parsed JSON data
            
        Raises:
            FileNotFoundError: If the file doesn't exist
            json.JSONDecodeError: If the file isn't valid JSON
            IOError: If there's an error reading the file
        """
        try:
            path = Path(file_path)
            logger.debug(f"Reading JSON file: {path}")
            with open(path, "r", encoding="utf-8") as f:
                data = json.load(f)
            return data
        except FileNotFoundError:
            logger.error(f"JSON file not found: {path}")
            raise
        except json.JSONDecodeError as e:
            logger.error(f"Invalid JSON in file {path}: {str(e)}")
            raise
        except IOError as e:
            logger.error(f"Error reading JSON file {path}: {str(e)}")
            raise
    
    @staticmethod
    def write_json(file_path: Union[str, Path], data: Dict[str, Any], indent: int = 2) -> None:
        """Write data to a JSON file.
        
        Args:
            file_path: Path to the JSON file to write
            data: Data to write to the file
            indent: Number of spaces for indentation
            
        Raises:
            TypeError: If the data isn't JSON-serializable
            IOError: If there's an error writing the file
        """
        try:
            path = Path(file_path)
            # Create directory if it doesn't exist
            path.parent.mkdir(parents=True, exist_ok=True)
            
            logger.debug(f"Writing JSON to file: {path}")
            with open(path, "w", encoding="utf-8") as f:
                json.dump(data, f, indent=indent)
        except TypeError as e:
            logger.error(f"Data is not JSON-serializable: {str(e)}")
            raise
        except IOError as e:
            logger.error(f"Error writing JSON to file {path}: {str(e)}")
            raise
    
    @staticmethod
    def ensure_directory(directory_path: Union[str, Path]) -> None:
        """Ensure a directory exists, creating it if necessary.
        
        Args:
            directory_path: Path to the directory
            
        Raises:
            IOError: If there's an error creating the directory
        """
        try:
            path = Path(directory_path)
            logger.debug(f"Ensuring directory exists: {path}")
            path.mkdir(parents=True, exist_ok=True)
        except IOError as e:
            logger.error(f"Error creating directory {path}: {str(e)}")
            raise
    
    @staticmethod
    def copy_file(source_path: Union[str, Path], dest_path: Union[str, Path]) -> None:
        """Copy a file from source to destination.
        
        Args:
            source_path: Path to the source file
            dest_path: Path to the destination file
            
        Raises:
            FileNotFoundError: If the source file doesn't exist
            IOError: If there's an error copying the file
        """
        try:
            src = Path(source_path)
            dst = Path(dest_path)
            # Create destination directory if it doesn't exist
            dst.parent.mkdir(parents=True, exist_ok=True)
            
            logger.debug(f"Copying file from {src} to {dst}")
            shutil.copy2(src, dst)
        except FileNotFoundError:
            logger.error(f"Source file not found: {src}")
            raise
        except IOError as e:
            logger.error(f"Error copying file from {src} to {dst}: {str(e)}")
            raise
    
    @staticmethod
    def delete_file(file_path: Union[str, Path]) -> None:
        """Delete a file.
        
        Args:
            file_path: Path to the file to delete
            
        Raises:
            FileNotFoundError: If the file doesn't exist
            IOError: If there's an error deleting the file
        """
        try:
            path = Path(file_path)
            logger.debug(f"Deleting file: {path}")
            path.unlink()
        except FileNotFoundError:
            logger.error(f"File not found for deletion: {path}")
            raise
        except IOError as e:
            logger.error(f"Error deleting file {path}: {str(e)}")
            raise
    
    @staticmethod
    def list_files(directory_path: Union[str, Path], pattern: str = "*") -> List[Path]:
        """List files in a directory matching a pattern.
        
        Args:
            directory_path: Path to the directory
            pattern: Glob pattern for matching files
            
        Returns:
            List of file paths
            
        Raises:
            FileNotFoundError: If the directory doesn't exist
            IOError: If there's an error listing files
        """
        try:
            path = Path(directory_path)
            logger.debug(f"Listing files in {path} matching pattern '{pattern}'")
            return list(path.glob(pattern))
        except FileNotFoundError:
            logger.error(f"Directory not found: {path}")
            raise
        except IOError as e:
            logger.error(f"Error listing files in {path}: {str(e)}")
            raise
    
    @staticmethod
    def list_directories(directory_path: Union[str, Path], pattern: str = "*") -> List[Path]:
        """List directories in a directory matching a pattern.
        
        Args:
            directory_path: Path to the directory
            pattern: Glob pattern for matching directories
            
        Returns:
            List of directory paths
            
        Raises:
            FileNotFoundError: If the directory doesn't exist
            IOError: If there's an error listing directories
        """
        try:
            path = Path(directory_path)
            logger.debug(f"Listing directories in {path} matching pattern '{pattern}'")
            # Get all items matching the pattern and filter for directories only
            return [p for p in path.glob(pattern) if p.is_dir()]
        except FileNotFoundError:
            logger.error(f"Directory not found: {path}")
            raise
        except IOError as e:
            logger.error(f"Error listing directories in {path}: {str(e)}")
            raise

File: javascript/src/utils/FileOperations.ts
import fs from 'fs';
import path from 'path';
import { logger } from './LoggerSetup';

/**
 * File operations utilities for reading and writing files.
 * Provides methods for working with files and directories,
 * including reading, writing, copying, and deleting files.
 */
export class FileOperations {
  /**
   * Read a file and return its contents
   * 
   * @param filePath - Path to the file to read
   * @param encoding - File encoding (default: utf-8)
   * @returns The file contents as a string
   * @throws Error if the file doesn't exist or can't be read
   */
  public static readFile(filePath: string, encoding: BufferEncoding = 'utf-8'): string {
    try {
      logger.debug(`Reading file: ${filePath}`);
      return fs.readFileSync(filePath, { encoding });
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        const errorMessage = `File not found: ${filePath}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      } else {
        const errorMessage = `Error reading file ${filePath}: ${error}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      }
    }
  }

  /**
   * Write content to a file
   * 
   * @param filePath - Path to the file to write
   * @param content - Content to write to the file
   * @param encoding - File encoding (default: utf-8)
   * @throws Error if there's an error writing the file
   */
  public static writeFile(filePath: string, content: string, encoding: BufferEncoding = 'utf-8'): void {
    try {
      // Create directory if it doesn't exist
      const dirname = path.dirname(filePath);
      if (!fs.existsSync(dirname)) {
        fs.mkdirSync(dirname, { recursive: true });
      }

      logger.debug(`Writing to file: ${filePath}`);
      fs.writeFileSync(filePath, content, { encoding });
    } catch (error) {
      const errorMessage = `Error writing to file ${filePath}: ${error}`;
      logger.error(errorMessage);
      throw new Error(errorMessage);
    }
  }

  /**
   * Append content to a file
   * 
   * @param filePath - Path to the file to append to
   * @param content - Content to append to the file
   * @param encoding - File encoding (default: utf-8)
   * @throws Error if there's an error appending to the file
   */
  public static appendToFile(filePath: string, content: string, encoding: BufferEncoding = 'utf-8'): void {
    try {
      // Create directory if it doesn't exist
      const dirname = path.dirname(filePath);
      if (!fs.existsSync(dirname)) {
        fs.mkdirSync(dirname, { recursive: true });
      }

      logger.debug(`Appending to file: ${filePath}`);
      fs.appendFileSync(filePath, content, { encoding });
    } catch (error) {
      const errorMessage = `Error appending to file ${filePath}: ${error}`;
      logger.error(errorMessage);
      throw new Error(errorMessage);
    }
  }

  /**
   * Read a JSON file and return its contents
   * 
   * @param filePath - Path to the JSON file to read
   * @returns The parsed JSON data
   * @throws Error if the file doesn't exist, isn't valid JSON, or can't be read
   */
  public static readJson<T = Record<string, unknown>>(filePath: string): T {
    try {
      logger.debug(`Reading JSON file: ${filePath}`);
      const content = fs.readFileSync(filePath, { encoding: 'utf-8' });
      return JSON.parse(content) as T;
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        const errorMessage = `JSON file not found: ${filePath}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      } else if (error instanceof SyntaxError) {
        const errorMessage = `Invalid JSON in file ${filePath}: ${error.message}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      } else {
        const errorMessage = `Error reading JSON file ${filePath}: ${error}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      }
    }
  }

  /**
   * Write data to a JSON file
   * 
   * @param filePath - Path to the JSON file to write
   * @param data - Data to write to the file
   * @param indent - Number of spaces for indentation (default: 2)
   * @throws Error if the data isn't JSON-serializable or there's an error writing the file
   */
  public static writeJson(filePath: string, data: unknown, indent = 2): void {
    try {
      // Create directory if it doesn't exist
      const dirname = path.dirname(filePath);
      if (!fs.existsSync(dirname)) {
        fs.mkdirSync(dirname, { recursive: true });
      }

      logger.debug(`Writing JSON to file: ${filePath}`);
      fs.writeFileSync(filePath, JSON.stringify(data, null, indent), 'utf-8');
    } catch (error) {
      const errorMessage = `Error writing JSON to file ${filePath}: ${error}`;
      logger.error(errorMessage);
      throw new Error(errorMessage);
    }
  }

  /**
   * Ensure a directory exists, creating it if necessary
   * 
   * @param directoryPath - Path to the directory
   * @throws Error if there's an error creating the directory
   */
  public static ensureDirectory(directoryPath: string): void {
    try {
      logger.debug(`Ensuring directory exists: ${directoryPath}`);
      if (!fs.existsSync(directoryPath)) {
        fs.mkdirSync(directoryPath, { recursive: true });
      }
    } catch (error) {
      const errorMessage = `Error creating directory ${directoryPath}: ${error}`;
      logger.error(errorMessage);
      throw new Error(errorMessage);
    }
  }

  /**
   * Copy a file from source to destination
   * 
   * @param sourcePath - Path to the source file
   * @param destPath - Path to the destination file
   * @throws Error if the source file doesn't exist or there's an error copying the file
   */
  public static copyFile(sourcePath: string, destPath: string): void {
    try {
      // Create destination directory if it doesn't exist
      const dirname = path.dirname(destPath);
      if (!fs.existsSync(dirname)) {
        fs.mkdirSync(dirname, { recursive: true });
      }

      logger.debug(`Copying file from ${sourcePath} to ${destPath}`);
      fs.copyFileSync(sourcePath, destPath);
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        const errorMessage = `Source file not found: ${sourcePath}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      } else {
        const errorMessage = `Error copying file from ${sourcePath} to ${destPath}: ${error}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      }
    }
  }

  /**
   * Delete a file
   * 
   * @param filePath - Path to the file to delete
   * @throws Error if the file doesn't exist or there's an error deleting the file
   */
  public static deleteFile(filePath: string): void {
    try {
      logger.debug(`Deleting file: ${filePath}`);
      fs.unlinkSync(filePath);
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        const errorMessage = `File not found for deletion: ${filePath}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      } else {
        const errorMessage = `Error deleting file ${filePath}: ${error}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      }
    }
  }

  /**
   * List files in a directory matching a pattern
   * 
   * @param directoryPath - Path to the directory
   * @param pattern - Regular expression pattern for matching files (default: all files)
   * @returns List of file paths
   * @throws Error if the directory doesn't exist or there's an error listing files
   */
  public static listFiles(directoryPath: string, pattern?: RegExp): string[] {
    try {
      logger.debug(`Listing files in ${directoryPath}${pattern ? ` matching pattern '${pattern}'` : ''}`);
      
      if (!fs.existsSync(directoryPath)) {
        const errorMessage = `Directory not found: ${directoryPath}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      }
      
      const files = fs.readdirSync(directoryPath);
      
      if (pattern) {
        return files
          .filter(file => pattern.test(file))
          .map(file => path.join(directoryPath, file));
      } else {
        return files.map(file => path.join(directoryPath, file));
      }
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
        const errorMessage = `Directory not found: ${directoryPath}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      } else if ((error as Error).message.includes('Directory not found')) {
        throw error;
      } else {
        const errorMessage = `Error listing files in ${directoryPath}: ${error}`;
        logger.error(errorMessage);
        throw new Error(errorMessage);
      }
    }
  }
}